rest/diagnostic_api.go

```diff
@@ -0,0 +1,222 @@
+	resp.AdminRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory)
+	return resp
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	username := mux.Vars(h.rq)["name"]
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(username))
+	if err != nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+	if user == nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+
+	authenticator := h.db.Authenticator(h.ctx())
+	if err != nil {
+		return err
+	}
+
+	resp := newGetAllChannelsResponse()
+
+	// Rebuild roles and channels
+	for _, dsName := range h.db.DataStoreNames() {
+		// skip if scope/collection has been removed
+		if _, ok := h.db.CollectionNames[dsName.Scope]; !ok {
+			continue
+		}
+		err = authenticator.RebuildCollectionChannels(user, dsName.Scope, dsName.Collection)
+		if err != nil {
+			return fmt.Errorf("Could not rebuild channels for %s: %w", dsName.String(), err)
+		}
+	}
+	if err := authenticator.RebuildRoles(user); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+	for roleName, roleEntry := range user.RoleNames() {
+		role, err := h.db.Authenticator(h.ctx()).GetRoleIncDeleted(roleName)
+		if role != nil && role.IsDeleted() {
+			base.InfofCtx(h.ctx(), base.KeyDiagnostic, "Role %s deleted, continuing")
+		}
+		if err != nil {
+			return err
+		}
+		if role == nil {
+			continue
+		}
+		for _, dsName := range h.db.DataStoreNames() {
+			keyspace := fmt.Sprintf("%s.%s", dsName.Scope, dsName.Collection)
+			collectionChannels := role.CollectionChannels(dsName.Scope, dsName.Collection)
+			channelHistory := role.CollectionChannelHistory(dsName.Scope, dsName.Collection)
+			if len(collectionChannels) == 0 && len(channelHistory) == 0 {
+				continue
+			}
+			for channel, chanEntry := range collectionChannels {
+				// loop over current role channels
+				if channel == channels.DocumentStarChannel {
+					continue
+				}
+				grantInfo := auth.GrantHistory{Entries: []auth.GrantHistorySequencePair{{StartSeq: chanEntry.VbSequence.Sequence}}, Source: chanEntry.Source}
+				if roleEntry.VbSequence.Sequence > chanEntry.VbSequence.Sequence {
+					grantInfo.Entries[len(grantInfo.Entries)-1].StartSeq = roleEntry.VbSequence.Sequence
+				}
+				resp.addRoleGrants(roleName, roleEntry.Source, keyspace, channel, grantInfo)
+				// loop over previous role channels
+			}
+			for channel, chanHistory := range channelHistory {
+				if roleEntry.VbSequence.Sequence > chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq {
+					chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq = roleEntry.VbSequence.Sequence
+				}
+				resp.addRoleGrants(roleName, roleEntry.Source, keyspace, channel, chanHistory)
+			}
+		}
+	}
+
+	for roleName, roleHist := range user.RoleHistory() {
+		role, err := h.db.Authenticator(h.ctx()).GetRole(roleName)
+		if err != nil {
+			return err
+		}
+		if role == nil {
+			continue
+		}
```

Comment:
Use the matching idiom from above:

```suggestion
		role, err := authenticator.GetRole(roleName)
		if err != nil {
			return fmt.Errorf("Error getting role %s: %w", roleName, err)
		}
		// deleted role will return nil with no error
		if role == nil {
			continue
		}
```